### ISSUE TITLE
Feature Request: Making Surrender button useful

### ISSUE THREAD

**Issue by Discontinuum** [wesnoth/wesnoth#6876]
### Describe the desired feature

So, right now Surrender button does nothing, just types "player surrendered" in the chat and makes him an observer. Other than that, it doesn't help to tell the game outcome. Although for different gamemodes resignation probably leads to different things, and in some modes like coops doing nothing when a player resigns is okay.

So my proposal is: make a new event type, "surrender", which is triggered when one of sides surrenders. Then, scenarios that have meaningful logic of surrendering will use it (like, in usual 1v1 it may be just deleting the leader of a surrendered side, so the other side won), scenarios that don't care about surrenders just don't use it and the ignoring behaviour is the same as it is now.

**Comment by gfgtdf** [comment]
This is not as easy as it sounds at first (although still possible) since surrendering is explicitly allowed off-turn and only the currently active player can mosify the gamestate. I always thought the main point of the the surrender button is to simplify bookkeeping of the win/loss in the server statistics at some point.

**Comment by Discontinuum** [comment]
@gfgtdf chat or changing a unit's name is also allowed offturn, but affects the game state

**Comment by ProditorMagnus** [comment]
Chat doesnt affect gamestate, and name is not oos-safe.

**Comment by gfgtdf** [comment]
As ravana said, chatting or renaming units doesn't cause OOS, deleting units can.

**Comment by Discontinuum** [comment]
Well, my idea here is that the Surrender button should end the game. No more playing will be after it was hit. So you don't really care about consistency of the resulting game state as long as both players will see that the player who surrendered lost and the remaining one won

**Comment by Discontinuum** [comment]
Btw, about this there is a problem
>  to simplify bookkeeping of the win/loss in the server statistics at some point.

The problem is that players are not responsible about this button, so they may press it just for quitting the game, even if it's temporary and they're gonna return. So using it for bookkeeping is possible but the results will have false positives.

**Comment by soliton-** [comment]
The surrender button should not end the game. That would only make sense in 1vs1. The point of this would be to allow scenario logic to decide if the game ends here or whatever else is desired.

If this is easier to implement when restricting surrender to the surrendering side's turn then that would be fine to do IMO.

**Comment by gfgtdf** [comment]
> The surrender button should not end the game. That would only make sense in 1vs1.

The logic when a game ends would be the same as it is when a side dies normally is already there, (usually when all sides are that are not "defeated" belong to the same team (being defeated is a flag that is set usually when all leader die)) And also customizable. So i'm not sure whether more customization (a new event) is needed. The main challenge here is really to make it (flagging a side as defeated) work off-turn without risking OOS.

**Comment by Discontinuum** [comment]
My primary goal was to make it easy to determine who won using the replay only. Right now you have no way without actually replaying a replay in Wesnoth to see whose leader died. And checking for "surrender" messages makes no help cause it doesn't mean the game ended

**Comment by Discontinuum** [comment]
> What happens when one side of a team surrenders and the other side wants to continue playing?

Then nothing happens, just as it is now. I thought about it only for 1v1

**Comment by gfgtdf** [comment]
> We can certainly record a surrendered flag but otherwise IMO you should simply be able to continue playing and possibly still win.

The game would continue  in this case, since not all defeated members are allied.

> Feel free to simply remove that challenge and allow surrender only when it is the side's turn. I don't think this is important.

Hmm i'd relly prefer to support it, othwise people will just quit when its not their turn.

**Comment by Discontinuum** [comment]
I think I originally suggested making an event that is implemented by each scenario as the scenario's author wishes, if it's needed. So it's up to the scenario author how to deal with it in games with more than 2 players.

**Comment by soliton-** [comment]
Regardless of a new surrender event I think we can do the following:
- Upon surrender record that for the side
- When deciding game end:
  - if all sides of a team surrendered consider that team defeated
  - if all sides of a team are defeated in the current logic consider the team defeated
  - if there is a mix of surrendered and "normally" defeated sides do **not** consider the team defeated

**Comment by CelticMinstrel** [comment]
> This is not as easy as it sounds at first (although still possible) since surrendering is explicitly allowed off-turn and only the currently active player can mosify the gamestate.

Is switching controller possible off-turn without sync errors? Maybe surrendering could be implemented as something like "set the controller to the surrendered controller" and the surrendered event is then "a side with the surrendered controller starts a turn".

**Comment by CelticMinstrel** [comment]
As far as I know, that's still triggered by the side that has control. In other words, it's certainly possible for the current side that has control to ask other sides for input, but I think it's not possible for a side that doesn't have control to arbitrarily do something.

**Comment by Discontinuum** [comment]
Yes, it should be triggered by the side that has control. But maybe it's possible to make through the mp server, so the surrendering player sends the signal, and the mp server tells other clients to activate the surrendering event

**Comment by soliton-** [comment]
Those other clients may be in the middle of something. If we want to implement the surrender event and keep allowing surrender off-turn then some kind of delayed execution like what @CelticMinstrel suggested sounds good.
